feat(fluentd): Allow option to only log router logs to elasticsearch

n0n0x opened the pull request:
This allows to use elasticsearch output plugin to only parse and log deis router logs
We use this together with the following custom filter:

```
<filter kubernetes.var.log.containers.deis-router-**.log>
  @type parser
  key_name log
 reserve_data true
  <parse>
    @type regexp
    expression /^[^ ]* - (?<app_name>[^ ]*) - (?<remote_addr>[^ ]*) - (?<remote_user>[^ ]*) - (?<status_code>[^ ]*) - "(?<method>\S+[^\"])(?: +(?<path>[^\"]*?)(?: +\S*)?)?" - (?<bytes_sent>[^ ]*) - "(?<referer>[^ ]*)" - "(?<agent>[^\"]*)\" - \"(?<server_name>[^ ]*)" - (?<upstream_addr>[^ ]*) - (?<http_host>[^ ]*) - (?<upstream_response_time>[^ ]*) - (?<request_time>[^ ]*)$/
    time_format %d/%b/%Y:%H:%M:%S %z
  </parse>
</filter>
```

This filter basically gets the router logs and parses it into different fields so it can be filtered in elasticsearch.

This is our fluentd section of our deis workflow YAML settings using this code:
```yaml
fluentd:
  daemon_environment:
    org: "n0n0x"
    docker_tag: "v2.14.15-santi"
    ELASTICSEARCH_HOST: "elasticsearch.url.com"
    ELASTICSEARCH_SCHEME: "http"
    ELASTICSEARCH_PORT: "80"
    ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY: "True"
    CUSTOM_FILTER_1: "<filter kubernetes.var.log.containers.deis-router-**.log>\n  @type
      parser\n  key_name log\n reserve_data true\n  <parse>\n    @type regexp\n
      \   expression /^[^ ]* - (?<app_name>[^ ]*) - (?<remote_addr>[^ ]*) -
      (?<remote_user>[^ ]*) - (?<status_code>[^ ]*) - \"(?<method>\\S+[^\\\"])(?:
      +(?<path>[^\\\"]*?)(?: +\\S*)?)?\" - (?<bytes_sent>[^ ]*) - \"(?<referer>[^
      ]*)\" - \"(?<agent>[^\\\"]*)\\\" - \\\"(?<server_name>[^ ]*)\" - (?<upstream_addr>[^
      ]*) - (?<http_host>[^ ]*) - (?<upstream_response_time>[^ ]*) - (?<request_time>[^
      ]*)$/\n    time_format %d/%b/%Y:%H:%M:%S %z\n  </parse>\n</filter>"
```

n0n0x committed:
feat(fluentd): Allow option to only log router logs to elasticsearch

Cryptophobia commented:
@n0n0x , does it continue the same behaviour as before when `ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY` is unset?

Cryptophobia commented on rootfs/fluentd/sbin/stores/stores:
```diff
@@ -1,7 +1,7 @@
 
 source /fluentd/sbin/stores/deis
-source /fluentd/sbin/stores/elastic_search
+[ -e "${ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY}" ] && source /fluentd/sbin/stores/elastic_search
```
What does this line change actually do? When I run this line, it gives me an error status code. Are you sure this actually does what it is supposed to?

Cryptophobia replied:
I should clarify... This gives me an error status code no matter if it is set/unset or if it is True or False. It always give me an error status code on my shell...

n0n0x replied:
Sorry about this, this was fixed in another branch I was using but didn't merge it to this branch! `-e` evaluates if a file exists. I foolishly mistook it as 'empty'. But actually 'empty' is `-z` (string is zero size).

I have fixed this

n0n0x replied:
> What does this line change actually do? When I run this line, it gives me an error status code. Are you sure this actually does what it is supposed to?

@Cryptophobia 

This is used so we  don't store anything else but the router logs in Elasticsearch. Otherwise,  a second `<store> @type elasticsearch </store>` would be set in `<match **>` and would cause that everything (and the router logs) be logged in elasticsearch.

When `ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY` is set, then a new `<match>` tag is created for the router logs, and they are forwarded to elasticsearch and deis plugins.

Then, when configuring the more broad `<match **>` tag (this file in question), we evaluate if `ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY` is set. If it is, then we don't  configure the elasticsearch plugin in this tag, because we just want the router logs.

If `ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY` is not set, and the `ELASTICSEARCH_` variables are configured, then there is no new `<match>` tag for the router logs, and the elasticsearch plugin is configured in the more broad and default `<match **>` tag. This is the old behavior.

Here is a vague pseudo-config on how the fluentd configuration would look like in each situation

### WITHOUT any `ELASTICSEARCH_*`

```xml
<match **>
  type copy
  <store>
     @type deis
  </store>
</match>
```

### WITH  `ELASTICSEARCH_*` vars set but no `ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY` (old behavior)

```xml
<match **>
  type copy
  <store>
    @type elasticsearch
  </store>
  <store>
     @type deis
  </store>
</match>
```

### WITH  `ELASTICSEARCH_*` vars set AND `ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY` set as well

```xml
<match kubernetes.var.log.containers.deis-router-**.log>
  type copy
  <store>
    @type elasticsearch 
  </store>
  <store>
    @type deis
  </store>
</match>
<match **> 
  type copy 
  <store>
     @type deis 
  </store>
</match>
```

Cryptophobia commented on rootfs/Dockerfile:
```diff
@@ -11,7 +11,7 @@ RUN apk add --update --virtual .build-deps sudo build-base ruby-dev \
     &&  rake --rakefile=/fluentd/deis-output/Rakefile build \
     &&  fluent-gem install --no-document fluent-plugin-kubernetes_metadata_filter -v 2.1.4 \
-    &&  fluent-gem install --no-document fluent-plugin-elasticsearch -v 2.12.0 \
+    &&  fluent-gem install --no-document fluent-plugin-elasticsearch -v 4.0.4 \
```
Very nice! :100: Hopefully this upgrade will work for all users... Anyway we could find all changelongs from 2.12 to 4.0.4 so we can see what has changed and if anything would break.

n0n0x replied:
https://github.com/uken/fluent-plugin-elasticsearch/blob/master/History.md

there's a changelog! They don't seem to list any kind of breaking changes or compatibility issues anywhere though

n0n0x reviewed (commented):
> @n0n0x , does it continue the same behaviour as before when `ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY` is unset?

It does now!

Cryptophobia reviewed (changes requested):
I think the `-z` and `-n` are still incorrect. Can you please fix them @n0n0x?

https://unix.stackexchange.com/questions/109625/shell-scripting-z-and-n-options-with-if

From my testing:
```
$ unset $ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY
$ [ -z "${ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY}" ] && echo 'hello'
hello

$ export $ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY=True
$ [ -z $ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY ] && echo 'hello'

$ unset $ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY
$ [ -n $ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY ] && echo 'hello'
hello

$ export $ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY=True
$ [ -n $ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY ] && echo 'hello'
hello

```

Shouldn't the `-z` be the same as `-n`. They should both be the same or am I missing something...? @kingdonb , can you share your experience with bash and whether or not these checks would work as intended. I do not want to merge this in and then later have problems across all logging...

n0n0x commented:
The line in question is this one:

n0n0x commented:
@Cryptophobia, In this case, we should use `-z` because we shouldn't load elasticsearch plugin in `<match **>` tag when `sbin/stores/stores` is called if `ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY` is set. Otherwise, we would log everything to elasticsearch and not just the logs which beats the purpose of this PR.

This is the workflow:

`/sbin/boot` gets executed and then this code runs first:

![image](https://user-images.githubusercontent.com/800750/75903239-955ce600-5e41-11ea-95b3-b7533c4ac983.png)

basically checking if `ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY` is set. If it is set, then proceeds to configure elasticsearch in a new `<match>` tag that filters deis-router logs and adds elasticsearch and deis plugin as output stores. This ensures that the router logs are forwarded and stored to elasticsearch and any other configured store plugin.

Then, it proceeds to the next chunk of code which basically configures the more broader fluentd `<match **>` tag that matches everything else:

![image](https://user-images.githubusercontent.com/800750/75903530-00a6b800-5e42-11ea-86f0-6628fad78cce.png)

In here, it executes `sbin/stores/stores` which would actually include elasticsearch again, and everything would be logged to elasticsearch, the router logs, the kubernetes logs, the pods logs, everything. 
Since the purpose of this feature is to log only the router logs to elasticsearch and not everything, then I had to figure out a way to not include elasticsearch in the broader `<match **>` tag if we only want to log the router logs. That is why, in `sbin/stores/stores` I added this:

![image](https://user-images.githubusercontent.com/800750/75903757-54190600-5e42-11ea-9c1b-ae260eaf6985.png)

which basically checks if `ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY` is not set. If it is not set, then we clearly want everything to be stored in elasticsearch, so we `source` the elasticsearch pugin in `<match **>`. If it is set, then we skip it, because we already added it in the previous `<match>` tag that matched the router logs.

I could configure that line the following way as well:
```
[ -z "${ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY}" ] && source /fluentd/sbin/stores/elastic_search || echo "ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY variable is set, skipping loading elasticsearch plugin"
```

So:
1) If elasticsearch variables are set, _AND_ `ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY` is _NOT_ set, then we load elasticsearch in the `<match **>` tag and we do not create a new `<match  kubernetes.var.log.containers.deis-router-**.log>` tag. This logs everything to elasticsearch. And this is the default and old behavior.

2) If elasticsearch variables are set, _AND_ `ELASTICSEARCH_DEIS_ROUTER_LOGS_ONLY` _IS SET_, then we create a new `<match  kubernetes.var.log.containers.deis-router-**.log>` which forwards router logs to elasticsearch and any other configured store plugin. AND we _DO NOT_ include elasticsearch in the `<match **>`. (this last part is the `-z` in question). This feature is compatible with the default and old behavior described above, but adds the option to just log router logs to elasticsearch instead of everything.

Cryptophobia commented:
Great work @n0n0x ! Thank you for explaining it. I was missing the part about what exactly `sbin/stores/stores` does and how the plugins are actually loaded and when we need to skip loading the elasticsearch plugin. Sounds like now this is non-breaking and backwards compatible. So let's merge it in!

Cryptophobia reviewed (approved):
LGTM!

Cryptophobia commented:
Awesome work. Thank you for merging this in and for following up so quickly on the PR.
